lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,502 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <modem/at_parser.h>
+
+#include "at_match.h"
+
+enum num_type {
+	I16,
+	U16,
+	I32,
+	U32,
+	I64,
+};
+
+static bool is_crlf(const char *str)
+{
+	return strlen(str) == 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static bool is_response(const char *str)
+{
+	int diff;
+	static const char * const toclip[] = {
+		"\r\nOK\r\n",
+		"\r\nERROR\r\n",
+		"\r\n+CME ERROR",
+		"\r\n+CMS ERROR"
+	};
+
+	for (size_t i = 0; i < 4; i++) {
+		diff = strncmp(str, toclip[i], strlen(toclip[i]));
+		if (!diff) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static bool lookahead_crlf_or_more(const char *str)
+{
+	return strlen(str) >= 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static bool lookahead_crlf_and_more(const char *str)
+{
+	return strlen(str) > 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static void at_token_set_empty(struct at_token *token, const char *start)
+{
+	token->start = start;
+	token->len = 0;
+	token->type = AT_TOKEN_TYPE_EMPTY;
+	token->variant = AT_TOKEN_VARIANT_NO_COMMA;
+}
+
+static bool is_subparam(const struct at_token *token)
+{
+	switch (token->type) {
+	case AT_TOKEN_TYPE_INT:
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_EMPTY:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool is_valid_at_token(const struct at_token *token)
+{
+	if (!token->start) {
+		return false;
+	}
+
+	size_t len = strlen(token->start);
+
+	switch (token->type) {
+	case AT_TOKEN_TYPE_CMD_TEST:
+	case AT_TOKEN_TYPE_CMD_READ:
+	case AT_TOKEN_TYPE_CMD_SET:
+	case AT_TOKEN_TYPE_NOTIF:
+	case AT_TOKEN_TYPE_INT:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_STRING:
+	case AT_TOKEN_TYPE_RESP:
+		return len != 0 && token->len != 0 && token->len <= len;
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+		return len != 0 && token->len <= len;
+	case AT_TOKEN_TYPE_EMPTY:
+		return token->len == 0;
+	default:
+		return false;
+	}
+}
+
+static void at_parser_rewind(struct at_parser *parser)
+{
+	parser->cursor = parser->at;
+	parser->count = 0;
+	memset(&parser->prev_token, 0, sizeof(struct at_token));
+	parser->is_next_empty = false;
+}
+
+int at_parser_init(struct at_parser *parser, const char *at)
+{
+	if (!parser || !at) {
+		return -EINVAL;
```

Comment:
Perhaps you mean the inverse?

Reply:
I usually use `EFAULT` for input that causes bad memory accesses (e.g. `NULL`) and `EINVAL` for input that is not `NULL` but otherwise incorrect (we do the same in `libmodem` too).